### PR TITLE
feat: support sp exit and bucket migrate

### PR DESCRIPTION
**Pull request by will-2012**
### Description

The feat-sp-exit-dev branch is the new dev branch to support sp exit and bucket migration.

#### Layers
image

#### Architecture
image

##### Adapt workflow
Get Create Bucket Approval: VMG free size weighted pick a family for the bucket, if not found, create a new and retry pick
Put Object: VMG free size weighted pick a gvg(in bucket family) for the object(replace p2p network),if not found, create a new and retry pick, replicate use the gvg’s secondary sp list.
##### Bucket Migrate Workflow
The bucket user should ask new primary sp for the migration bucket, and get the approval of the new primary SP.
The bucket user submits a MigrationBucket transaction with the approval of the new primary SP.
SP Side:
1.BlockSyncer subscribe event from chain and build to the index(Index builder).
2.BucketMigrateScheduler query bucket migrate event from Metadata(Index querier).
3.BucketMigrateScheduler produces migrate execute plan(including many virtual groups). Each gvg pick new dest gvg.
4.BucketMigrateScheduler push gvg migrate meta to task dispatcher.
5. Executor fetches the migrate gvg task to execute, and period report progress.
6.BucketMigrateScheduler sends complete tx if all the gvg tasks are completed in the execute plan, which need to get secondary sp bls signature filled to transaction field.
##### SP Exit Workflow
The SP is self-applying for exit by sending the StorageProviderExit transaction to the blockchain.
SP Side:
Src SP:
1.BlockSyncer subscribe event from chain and build to the index(Index builder).
2.SPExitScheduler query sp exit event from Metadata(Index querier).
3.SPExitScheduler produce a swap out plan(each includes many virtual groups).
4.SPExitScheduler pick dest sp/get swap out approval and send swap out start tx to the chain.
5.SPExitScheduler dispatch swap out info dest sp.
6.SPExitScheduler query swap out complete event from Metadata, if all swap out are completed, and send sp exit completed tx.
Dest SP:
1.SPExitScheduler receives src sp get to swap out approval.
2.SPExitScheduler receives src sp swap out unit.
3. expand swap out to gvg list.
4. push gvg to task dispatcher, if all gvg are completed in swap out unit, send swap out complete tx.
5. Executor fetches the migrate gvg task to execute, and period report progress.

### Example

N/A.

### Changes

Notable changes: 
* N/A.

### CODE REVIEW

File: modular/executor/migrate_task.go [hunk 1]
```diff
@@ -126,74 +120,109 @@ func (e *ExecuteModular) doMigrationGVGTask(ctx context.Context, task coretask.M
...
-// 		log.CtxErrorw(ctx, "failed to migrate ec pieces", "error", err)
-// 	}
-// }
+func (e *ExecuteModular) checkGvgConflict(ctx context.Context, srcGvg, destGvg *virtualgrouptypes.GlobalVirtualGroup,
+	objectInfo *storagetypes.ObjectInfo, params *storagetypes.Params) error {
+	index := util.ContainOnlyOneDifferentElement(srcGvg.GetSecondarySpIds(), destGvg.GetSecondarySpIds())
+	if index == -1 {
+		return fmt.Errorf("invalid gvg secondary sp id list")
+	}
+	// srcSecondarySPID := srcGvg.GetSecondarySpIds()[index]
```
Comment (Fynnss) [inline modular/executor/migrate_task.go:129]: remove useless code

File: modular/executor/migrate_task.go [hunk 2]
```diff
@@ -126,74 +120,109 @@ func (e *ExecuteModular) doMigrationGVGTask(ctx context.Context, task coretask.M
...
+		log.CtxErrorw(ctx, "failed to done replicate piece", "dest_sp_endpoint", destSPEndpoint,
+			"segment_idx", segmentIdx, "error", err)
+		return err
+	}
+	if int(segmentIdx+1) >= len(objectInfo.GetChecksums()) {
+		log.CtxErrorw(ctx, "failed to done replicate piece, replicate idx out of bounds", "segment_idx", segmentIdx)
+		return ErrReplicateIdsOutOfBounds
+	}
+	// var blsPubKey bls.PublicKey
+	// err = veritySignature(ctx, rTask.GetObjectInfo().Id.Uint64(), rTask.GetGlobalVirtualGroupId(), integrity,
```
Comment (Fynnss) [inline modular/executor/migrate_task.go:215]: why comment `veritySignature`?

File: cmd/storage_provider/main.go [hunk 3]
```diff
@@ -116,6 +116,8 @@ func init() {
 		// recovery commands
 		command.RecoverObjectCmd,
 		command.RecoverPieceCmd,
+		// sp exit and migrate
```
[the remark this answers is not in the document]
Comment (will-2012) [reply cmd/storage_provider/main.go]: Fixed.

File: base/gfspvgmgr/virtual_group_manager.go [hunk 4]
```diff
@@ -393,7 +376,8 @@ func (vgm *virtualGroupManager) PickMigrateDestGlobalVirtualGroup(vgfID uint32)
 // if pick func returns ErrStaledMetadata, the caller need force refresh from metadata and retry pick.
 func (vgm *virtualGroupManager) ForceRefreshMeta() error {
 	// sleep 2 seconds for waiting a new block
-	time.Sleep(2 * time.Second)
+	// TODO: pre-allocator improve it.
```
[the remark this answers is not in the document]
Comment (will-2012) [reply base/gfspvgmgr/virtual_group_manager.go]: Fixed.

File: base/gnfd/gnfd_service.go [hunk 5]
```diff
@@ -249,32 +249,36 @@ func (g *Gnfd) ListBondedValidators(ctx context.Context) (validators []stakingty
...
 // ListVirtualGroupFamilies return the list of virtual group family.
+// TODO: improve it by metadata
 func (g *Gnfd) ListVirtualGroupFamilies(ctx context.Context, spID uint32) ([]*virtualgrouptypes.GlobalVirtualGroupFamily, error) {
 	startTime := time.Now()
 	defer metrics.GnfdChainTime.WithLabelValues("list_virtual_group_family").Observe(time.Since(startTime).Seconds())
 	client := g.getCurrentClient().GnfdClient()
 	var vgfs []*virtualgrouptypes.GlobalVirtualGroupFamily
 	resp, err := client.VirtualGroupQueryClient.GlobalVirtualGroupFamilies(ctx, &virtualgrouptypes.QueryGlobalVirtualGroupFamiliesRequest{
-		StorageProviderId: spID,
+		Pagination: &query.PageRequest{Limit: 10000000},
```
[the remark this answers is not in the document]
Comment (will-2012) [reply base/gnfd/gnfd_service.go]: Fixed.

File: modular/executor/migrate_task.go (hunk 1 above)
Comment (will-2012) [reply modular/executor/migrate_task.go]: Fixed.

File: modular/executor/migrate_task.go [hunk 6]
```diff
@@ -85,24 +79,25 @@ func (e *ExecuteModular) doMigrationGVGTask(ctx context.Context, task coretask.M
...
 		log.CtxErrorw(ctx, "failed to query bucket info by bucket name", "bucketName", bucketName, "error", err)
 		return err
 	}
+	// bucket migration, check secondary whether is conflict, if true replicate own secondary SP data to another SP
+	if bucketID != 0 {
+		if err = e.checkGvgConflict(ctx, task.GetSrcGvg(), task.GetDestGvg(), object.GetObjectInfo(), params); err != nil {
+			log.Debugw("no gvg conflict", "error", err)
+		}
+	}
 	if bucketID != bucketInfo.Id.Uint64() {
```
[the remark this answers is not in the document]
Comment (sysvm) [reply modular/executor/migrate_task.go:96]: fixed

**Review by sysvm — APPROVED** [review]
LGTM

**Pull request merged**